proof/refine/ARM/Finalise_R.thy

```diff
@@ -4016,6 +4013,81 @@ lemma unbind_maybe_notification_corres:
+    apply (rule corres_symb_exec_l)
+       apply (rule corres_guard_imp)
+         apply (rule_tac P="\<lambda>s. kheap s ptr = Some (kernel_object.SchedContext sc (objBits sc' - minSchedContextBits))"
+                     and P'="ko_at' sc' ptr"  in corres_inst)
+         apply (rule_tac n="objBits sc' - minSchedContextBits" in setSchedContext_no_stack_update_corres)
+            apply simp+
+      apply (wpsimp wp: get_sched_context_exs_valid simp: is_sc_obj_def obj_at_def)
+       apply (rename_tac ko; case_tac ko; simp)
+      apply simp
+     apply (wpsimp simp: obj_at_def is_sc_obj_def)+
+    apply (wpsimp wp: get_sched_context_no_fail)
+   apply (clarsimp simp: obj_at_def is_sc_obj_def)
+  apply simp
+  done
+
+lemma sched_context_maybe_unbind_ntfn_corres:
+  "corres dc
+     (invs and ntfn_at ntfn_ptr)
+     (invs' and ntfn_at' ntfn_ptr)
+     (sched_context_maybe_unbind_ntfn ntfn_ptr)
+     (schedContextMaybeUnbindNtfn ntfn_ptr)"
+  apply (clarsimp simp: sched_context_maybe_unbind_ntfn_def schedContextMaybeUnbindNtfn_def)
+  apply (clarsimp simp: maybeM_def get_sk_obj_ref_def liftM_def)
+  apply add_sym_refs
+  apply (rule corres_guard_imp)
+    apply (rule corres_split[OF _ get_ntfn_corres])
+      apply (rename_tac ntfnA ntfnH)
+      apply (rule corres_option_split)
+        apply (simp add: ntfn_relation_def)
+       apply (rule corres_return_trivial)
+      apply (rename_tac scAPtr)
+      apply (clarsimp simp: schedContextUnbindNtfn_def update_sk_obj_ref_def bind_assoc)
+      apply (rule corres_stateAssert_add_assertion[where P'=\<top>, simplified])
+       apply (rule_tac P="invs and ntfn_at ntfn_ptr and ko_at (Notification ntfnA) ntfn_ptr"
+                and P'="valid_objs' and ko_at' ntfnH ntfn_ptr and sym_refs_asrt"
+                and Q'1=\<top>
+                in corres_symb_exec_r'[THEN corres_guard_imp])
+            apply (rule_tac F="scNtfn rv = Some ntfn_ptr" in corres_gen_asm2)
+            apply clarsimp
+            apply (rule corres_split[OF _ get_ntfn_corres])
+              apply (rule corres_split[OF _ set_ntfn_corres])
+                 apply (rule_tac f'="scNtfn_update (\<lambda>_. None)" in update_sched_context_corres)
+                    apply (clarsimp simp: sc_relation_def objBits_def objBitsKO_def)+
+                apply (clarsimp simp: ntfn_relation_def split: Structures_A.ntfn.splits)
+               apply wpsimp+
+        apply (frule invs_valid_objs)
+        apply (frule (1) valid_objs_ko_at)
+        apply (clarsimp simp: invs_psp_aligned valid_obj_def valid_ntfn_def)
+       apply (clarsimp simp: valid_ntfn'_def ntfn_relation_def split: option.splits)
+       apply (drule_tac s="Some scAPtr" in sym)
+       apply (clarsimp simp: valid_ntfn'_def ntfn_relation_def sym_refs_asrt_def)
+       apply (frule (3) ntfnSc_sym_refsD[OF ko_at_obj_at', simplified])
+       apply normalise_obj_at'
+      apply (drule state_refs_of_cross_eq)
```

Comment:
I'd be surprised if this `drule state_refs_of_cross_eq` were necessary. That should be part of `add_sym_refs`

Reply:
`add_sym_refs` doesn't work here, and even though I applied it earlier, I didn't get the fact I needed which I got from `state_refs_of_cross_eq`

Reply:
So add_sym_refs does not add sym_refs to the concrete guard? It should be able to handle the fact that you’ve got invs in the abstract guard. Could you check?

Reply:
Sorted the schematic issues now, add_sym_refs work and we're not using this drule anymore

Reply:
Perfection